cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillController.java

```diff
@@ -484,6 +482,120 @@ private void deletePVCs() {
     return resultFuture;
   }
 
+  /**
+   * Cleans up the resources for deployment and stateful set and returns the future.
+   */
+  private CompletableFuture<KubeTwillController> cleanupResources() {
+    CompletableFuture<KubeTwillController> resultFuture;
+    CompletionStage<String> completionStage;
+    resultFuture = new CompletableFuture<>();
+    // delete the resource
+    completionStage = deleteResource();
+    completionStage.whenComplete((name, t) -> {
+      if (t != null) {
+        resultFuture.completeExceptionally(t);
+      } else {
+        resultFuture.complete(KubeTwillController.this);
+      }
+    });
+    // If the deletion was successful, reflect the status into the completion future
+    // If the deletion was failed, don't change the completion state so that caller can retry
+    resultFuture.thenAccept(completion::complete);
+    return resultFuture;
+  }
+
+  /**
+   * Cleans up the job and returns future with correct job status.
+   */
+  private CompletableFuture<KubeTwillController> cleanupJob() {
+    // get job status
+    CompletionStage<String> completionStage = getJobCompletionStatus();
+    completionStage.whenComplete((name, t) -> {
+      if (t != null) {
+        completion.completeExceptionally(t);
+      } else {
+        completion.complete(KubeTwillController.this);
+      }
+    });
+    return completion;
+  }
+
+  /**
+   * Returns job completion status and attempts to delete the job.
+   */
+  private CompletionStage<String> getJobCompletionStatus() {
```

Comment:
nit: good to rename to indicate the job will be deleted by this method. something like deleteJobAndGetCompletionStatus()